src/java/org/apache/cassandra/db/marshal/VectorType.java

```diff
@@ -18,174 +18,612 @@
+            {
+                V bb = accessor.slice(buffer, offset, elementLength);
+                offset += elementLength;
+                elementSerializer.validate(bb, accessor);
+                result.add(bb);
+            }
+            if (!accessor.isEmptyFromOffset(buffer, offset))
+                throw new MarshalException("Unexpected extraneous bytes after list value");
+
+            return result;
         }
 
         @Override
-        public ByteBuffer serialize(float[] value)
+        public <V> V serializeRaw(List<V> value, ValueAccessor<V> accessor)
         {
-            return getByteBuffer(value);
+            if (value == null)
+                return accessor.empty();
+            check(value);
+
+            int size = elementType.valueLengthIfFixed();
+            V bb = accessor.allocate(size * dimension);
+            int position = 0;
+            for (V v : value)
+                position += accessor.copyTo(v, 0, bb, accessor, position, size);
+            return bb;
         }
 
-        public static ByteBuffer getByteBuffer(float[] value)
+        @Override
+        public ByteBuffer serialize(List<T> value)
         {
             if (value == null)
                 return ByteBufferUtil.EMPTY_BYTE_BUFFER;
+            check(value);
 
-            var bb = ByteBuffer.allocate(4 * value.length);
-            for (var v : value)
-                bb.putFloat(v);
-            bb.position(0);
+            ByteBuffer bb = ByteBuffer.allocate(elementType.valueLengthIfFixed() * dimension);
+            for (T v : value)
+                bb.put(elementSerializer.serialize(v).duplicate());
+            bb.flip();
             return bb;
         }
 
         @Override
-        public <V> float[] deserialize(V value, ValueAccessor<V> accessor)
+        public <V> List<T> deserialize(V input, ValueAccessor<V> accessor)
         {
-            if (value == null || accessor.isEmpty(value))
+            if (isNull(input, accessor))
                 return null;
+            List<T> result = new ArrayList<>(dimension);
+            int offset = 0;
+            int elementLength = elementType.valueLengthIfFixed();
+            for (int i = 0; i < dimension; i++)
+            {
+                V bb = accessor.slice(input, offset, elementLength);
+                offset += elementLength;
+                elementSerializer.validate(bb, accessor);
+                result.add(elementSerializer.deserialize(bb, accessor));
+            }
+            if (!accessor.isEmptyFromOffset(input, offset))
+                throw new MarshalException("Unexpected extraneous bytes after list value");
 
-            var vector = new float[dimensions];
-            for (int i = 0, offset = 0; i < dimensions; offset += 4, i++)
-                vector[i] = accessor.getFloat(value, offset);
-            return vector;
+            return result;
+        }
+
+        public <V> float[] deserializeFloatArray(V input, ValueAccessor<V> accessor)
```

Comment:
This was added by me to try and avoid creating a `List<Float>` and the associated boxing when we only want a `float[]`.